### PR TITLE
Refactoring to use a HookManager class

### LINKED ISSUE #22: API usage improvements
The current API still feels a little verbose. I'd really like to offer something that is chainable, offering all the current functionality but a little more concise:

```js
const sayHi = hooks(name => `Hi ${name}`, [ hook1 ]).params('name').defaults({
  name: 'Dave'
});
```

This would still work on objects:

```js
const o = {
  sayHi(name) { return `Hi ${name}` }
}

hooks(o, {
  sayHi: hooks([ hook1 ]).params('name').defaults({
    name: 'Dave'
  })
});
```

### LINKED ISSUE #28: Performance optimizations
One of the beta testers reported a noticeable performance difference in the latest version vs the current Feathers hooks. I ran a simple benchmark with different configurations and got the following results:

```
Baseline 11
Empty hooks 115
Single simple hook 138
Single hook and withParams 813
Single hook, multiple withParams 1331
```

Using the following test code:

```js
  it.only('benchmark', async () => {
    const getRuntime = async (callback: () => Promise<any>) => {
      const start = Date.now();
    
      for (let i = 0; i < 100000; i++) {
        await callback();
      }
    
      return Date.now() - start;
    }
    const hello = async (name: string, _params: any = {}) => {
      return `Hello ${name}`;
    };
    console.log('Baseline', await getRuntime(() => hello('Dave')));

    const hookHello1 = hooks(hello, []);
    console.log('Empty hooks', await getRuntime(() => hookHello1('Dave')));

    const hookHello2 = hooks(hello, [
      async (_ctx: HookContext, next: NextFunction) => {
        await next();
      }
    ]);

    console.log('Single simple hook', await getRuntime(() => hookHello2('Dave')));

    const hookHello3 = hooks(hello, {
      middleware: [
        async (_ctx: HookContext, next: NextFunction) => {
          await next();
        }
      ],
      context: [withParams('name')]
    });

    console.log('Single hook and withParams', await getRuntime(() => hookHello3('Dave')));


    const hookHello4 = hooks(hello, {
      middleware: [
        async (_ctx: HookContext, next: NextFunction) => {
          await next();
        }
      ],
      context: [withParams(), withParams('name'), withParams()]
    });

    console.log('Single hook, multiple withParams', await getRuntime(() => hookHello4('Dave')));
  });
```

While there will be always some kind of overhead of course, I think runtime could be improved by wrapping the context updaters into a single function instead of looping every time.

### PR DESCRIPTION
**Pull request by daffl**
This is a refactoring that addresses performance and API improvements.

- Closes #22 
- Closes #28
- Closes #29

### CODE REVIEW

File: packages/hooks/src/index.ts [hunk 1]
```diff
@@ -1,32 +1,42 @@
...
 export function hooks<T = any> (
-  hooks?: HookSettings
+  _manager?: HookOptions
 ): any;
 // Fallthrough to actual implementation
 export function hooks (...args: any[]) {
   const [ target, _hooks ] = args;
 
-  if (typeof target === 'function' && Array.isArray(_hooks.middleware || _hooks)) {
+  if (typeof target === 'function' && _hooks instanceof HookManager) {
```
Comment (bertho-zero) [inline packages/hooks/src/index.ts]:
```suggestion
  if (typeof target === 'function' && (_hooks instanceof HookManager || Array.isArray(_hooks))) {
```

In tests there is not one that accepts an array of hooks, without being wrapped beforehand:

```js
const sayHi = hooks(hello, [
  async (context, next) => {
    await next();
    context.result += '!';
  }
]);
```

**Review by bertho-zero — CHANGES_REQUESTED** [review]
We can also add a test for that:

```js
console.log(customContext.hasOwnProperty('message')); // true
console.log(customContext.hasOwnProperty('arguments')); // true
console.log(customContext.hasOwnProperty('name')); // false
console.log(Object.keys(customContext)); // [ 'message', 'arguments', 'result' ]
console.log(JSON.stringify(customContext)); // {"message":"Custom message","arguments":["Changed",{}],"result":"Hello Changed"}
```

There is always a problem with the named parameters, We cannot copy getter/setter properties from the prototype to the instance. No worries about the props.

Stackoverflow: https://stackoverflow.com/questions/6039676/copying-javascript-getters-setters-to-another-prototype-object

UPDATE: Works fine:

```js
for (const key in ctx) {
  const getter = ctx.constructor.prototype.__lookupGetter__(key);

  if (getter) {
    ctx.__defineGetter__(key, getter);
    ctx.__defineSetter__(key, ctx.constructor.prototype.__lookupSetter__(key));
  } else {
    ctx[key] = ctx[key];
  }
}
```

**Review by bertho-zero — APPROVED** [review]

**Comment by bertho-zero** [comment]
Note: wrapping an array of hooks with a `middleware` call is only useful if you chain with `.params` or `.props`

**Comment by daffl** [comment]
This has been published as `v0.5.0-alpha.0`. I think it's still missing default parameter initialization but wondering if it could just be combined with `.props(object|initFunction)`.

**Pull request merged**